compiler/qsc_linter/src/linter/ast.rs

```diff
@@ -106,25 +152,20 @@ macro_rules! declare_ast_lints {
 
     // Declare & implement a struct representing a lint.
     (@LINT_STRUCT $lint_name:ident, $default_level:expr, $msg:expr, $help:expr) => {
+
+        #[derive(Default)]
         pub(crate) struct $lint_name {
             level: LintLevel,
         }
 
-        impl Default for $lint_name {
-            fn default() -> Self {
-                Self { level: Self::DEFAULT_LEVEL }
-            }
-        }
-
-        impl From<LintLevel> for $lint_name {
-            fn from(value: LintLevel) -> Self {
-                Self { level: value }
-            }
-        }
-
         impl $lint_name {
             const DEFAULT_LEVEL: LintLevel = $default_level;
 
+            fn new() -> Self {
+                #[allow(clippy::needless_update)]
+                Self { level: Self::DEFAULT_LEVEL, ..Default::default() }
+            }
```

Comment:
It looks like this and some of the other changes to this file may be left over from a different design... the lint is warning that the only element of the struct has already been set, which is true. So the existing `default` implementation was correct and can be restored, this `new` can be removed and places where it was called returned to calling `default`.

Reply:
The design for the `default` doesn't work for when we want to have lints that have state. 

It is better, and less confusing, imho that the design for HIR lints and AST lints match so that devs don't need to memorize two different ways of making lints. Having feature parity earlier would have prevented certain lints from being biased towards being implemented as HIR lints because of the added flexibility instead of what would probably have been more appropriately AST lints.

Reply:
As such, in this PR, instead of making use of the more flexibly design of HIR lints, I am adding these features to the AST lint design and making AST lints to make the AST and HIR lints have feature parity.

Reply:
Oh, I see what you mean. With your latest changes now there is an AST lint that has more state; in the earlier version this change wasn't needed but now it's necessary. Thanks for clarifying!